src/views/Routes.vue

```diff
@@ -110,6 +121,64 @@ import RouteList from '@/components/RouteList.vue';
 import Spinner from '@/components/Spinner.vue';
 import UserLoader from '@/mixins/UserLoader';
 
+const secMs = 1000;
+const hourMs = 3600 * secMs;
+const dayMs = 24 * hourMs;
+
+// These functions are exported so they can be exercised by unit tests. Ideally
+// we could use a module that lets tests access them without exporting them,
+// but:
+// - The 'rewire' module is unable to import .vue files.
+// - The 'babel-plugin-rewire' module doesn't seem to work with Jest:
+//   https://github.com/facebook/create-react-app/issues/5662.
+
+// Formats the supplied duration as a string, e.g. '1 day, 3 hours' or '4 hours,
+// 23 minutes, 2 seconds'.
+// - If longer than a week, we display it with day granularity.
+// - If longer than a day, we display it with hour granularity.
+// - Otherwise, we display it with second granularity.
+export function formatDuration(ms: number): string {
+  let units: humanizeDuration.Unit[];
+  if (ms > 7 * dayMs) {
+    units = ['w', 'd'];
+  } else if (ms > dayMs) {
+    units = ['d', 'h'];
+  } else {
+    units = ['h', 'm', 's'];
+  }
+  return humanizeDuration(ms, { units, round: true });
+}
+
+// Gets the next timeout for updating the competition time message, given
+// |remainMs| until the upcoming event (i.e. competition start or end).
+export function getNextTimeout(remainMs: number): number {
+  if (remainMs <= 0) throw new Error(`Invalid duration ${remainMs}`);
+
+  // Computes the delay to return.
+  // |granularityMs| is the granularity used to display durations.
+  // |thresholdMs| is the duration at which we switch to a lower granularity.
+  const f = (granularityMs: number, thresholdMs: number) => {
+    if (remainMs <= thresholdMs) {
+      throw new Error(`Remaining time ${remainMs} <= ${thresholdMs}`);
+    }
+    // Compute the time until the next update.
+    let delayMs = remainMs % granularityMs;
```

Comment:
Should we assert that granularityMs is not 0?

Reply:
leaving as-is since this is part of the implementation of getNextTimeout, and granularityMs isn't data that's being passed in from outside